### DISCUSSION TITLE
openssl 3.2.1: on ios the integrity check for fips provider failing

### DISCUSSION THREAD

**Discussion by danbey** [openssl/openssl#24830] [Q&A]
Update: I manage to load the fips on ios only when hard-coded passing the integrity check libcrypto is doing for fips.dylib. Looks like there is a bug when creating/parsing the checksum from the fips.dylib

Hi, 

1. I have this error when trying to load fips with OSSL_PROVIDER_load. This is my code:
    OSSL_PROVIDER *fips = OSSL_PROVIDER_load(NULL, "fips");
    if (fips == NULL) {
            printf("Failed to load FIPS provider\n");
    }

    This is the errors i'm getting from openssl log:
    OpenSSL Error: error:1C8000D5:Provider routines::missing config data
    OpenSSL Error: error:1C8000E0:Provider routines::fips module entering error state
    OpenSSL Error: error:1C8000D8:Provider routines::self test post failure
    OpenSSL Error: error:078C0105:common libcrypto routines::init fail
   

2. In my app code i'm setting on runtime the OPENSSL_CONF and the OPENSSL_MODULES:
   NSString *fipsdylibPath = [[NSBundle mainBundle] pathForResource:@"fips" ofType:@"dylib"];
   NSString *libPath = [fipsdylibPath stringByDeletingLastPathComponent];
   if (libPath) {
        setenv("OPENSSL_MODULES", [libPath UTF8String], 1);

3. In my app code i'd tried to load the fips and call a function that is in fips.dylib, this was succeeded. This my code for it:
  NSString *libraryPath = [[NSBundle mainBundle] pathForResource:@"fips" ofType:@"dylib"];
  if (libraryPath) {
         void *handle = dlopen([libraryPath UTF8String], RTLD_NOW);

         if (handle) {
             void (*exampleFunction)() = dlsym(handle, "ossl_provider_module_name");

             if (exampleFunction) {
                 // Call the function
                 exampleFunction();
                 NSLog(@"Function called successfully");
             } else {
                 NSString *errorMessage = [NSString stringWithFormat:@"Symbol not found: %s", dlerror()];
                 NSLog(@"%@", errorMessage);
             }

             dlclose(handle);
         } else {
             NSString *errorMessage = [NSString stringWithFormat:@"Library not loaded: %s", dlerror()];
             NSLog(@"%@", errorMessage);
         }
     } else {
         NSString *errorMessage = @"Library path not found";
         NSLog(@"%@", errorMessage);
     }

5. I add a code that prints al the loaded provider to see which provider has been loaded. this is my code for it:
    OSSL_PROVIDER_do_all(NULL, (provider_clk)provider_callback, NULL);

// Callback function to check each provider's name
static int provider_callback(OSSL_PROVIDER *provider, void *arg)
{
    const char *provider_name = OSSL_PROVIDER_get0_name(provider);

    fprintf(stdout, "provider name : %s\n",  provider_name ? provider_name : "unknown");
    return 1;
}

6. When playing with the openssl.cnf to load only default provider or only the base+fips, i can see in the printing that it is loaded (one time the default and other only the base without fips ) . This is the 2 option of my .cnf : 
default option:
    [openssl_init]
    providers = provider_sect

    [provider_sect]
    default = default_sect

    [default_sect]
    activate = 1

    [base_sect]
    activate = 1

fips option:
    [openssl_init]
    providers = provider_sect

    [provider_sect]
    fips = fips_sect
    base = base_sect

    [fips_sect]
    activate = 1
    conditional-errors = 1
    security-checks = 1
    tls1-prf-ems-check = 1
    drbg-no-trunc-md = 1
    module-mac = HEXNUM

    [default_sect]
    activate = 1

    [base_sect]
    activate = 1

7. I tried loading the legacy.dylib and succeeded, i can see in my print the legacy is loaded. This is the code that load the legacy:
           OSSL_PROVIDER *fips = OSSL_PROVIDER_load(NULL, "legacy");

Thanks,

**Comment by pgarba** [comment]
I think the FIPS self test is failing because after you sign the FIPS provider dylib the fingerprint will change and loading of the FIPS provider will abort. You need to find a way to update the module-mac in the cnf file after the dylib was signed.
